### ISSUE TITLE
Rectangle has only 3 LatLngs if first and last latlng are equal

### ISSUE THREAD

**Issue by Falke-Design** [Leaflet/Leaflet#7464]
The logic of `_convertLatLngs` in L.Polygon removes the last latlng if it is the same as the first latlng. But a rectangle calls this too and then it results in having only 3 instead of 4 latlngs. --> Is not longer a valid Rectangle.

Both times the rectangle has only 3 latlngs left:
```
r = L.rectangle([[0,0],[0,0]])
console.log(r.getLatLngs())
r.setLatLngs([[1,1],[0,0],[0,0],[1,1]])
console.log(r.getLatLngs())
```

The problem code:
https://github.com/Leaflet/Leaflet/blob/620bc3e9c5d282090f429ac21d28d00fbffdb3db/src/layer/vector/Polygon.js#L99-L108

I think this can be easily fixed when this code is added to L.Rectangle:
```
 _convertLatLngs: function (latlngs) { 
 	var result = Polyline.prototype._convertLatLngs.call(this, latlngs), 
 	    len = result.length; 
  
 	// remove last point if it equals first one 
 	if (len >= 4 && result[0] instanceof LatLng && result[0].equals(result[len - 1])) { 
 		result.pop(); 
 	} 
 	return result; 
 }, 
```

Or we remove the check and the `pop` completely.

I will create a PR when I know which solution is preferred.

**Comment by johnd0e** [comment]
> having only 3 instead of 4 latlngs. --> Is not longer a valid Rectangle.

But we have no real definition of Rectangle in docs: https://leafletjs.com/reference-1.7.1.html#rectangle, so..

I see the issue in such way: Leaflet support of rectangle is rather limited.
As for me - LatLngs is just not appropriate data to represent rectangle, and besides of mentioned - there are a numerous other edge cases, where you will get in trouble trying operating of rectangle's latlngs directly.
May be we even should replace Rectangle's `get/setLatLngs` methods with some stubs, to prevent misusing.

But what if you use `get/setBounds` instead? Are there any case which is not covered with that methods?

**Comment by Falke-Design** [comment]
Yes, there is no definition for rectangles maybe we can change this. But how should this definition should look like?

- Minimum / Maximum 4 LatLngs?
- Are rectangles always horizontal? This would make problems with some Plugins. For example with rotation.
- Default creation over Bounds `L.rectangle([[0,0],[1,1]])`
- ...


I don't know if it would be a good idea to implement some validations to the `setLatLngs()` method. It depends what you want to validate against the misusing. For example I need the `.setLatLngs()` function to rotate a rectangle, so it would not possible for me to use `setBounds()`.
But it could make sense, to allow only 4 latlngs and / or don't remove the last latlng if it is the same as the first one (like I suggested)

With `setBounds` it is the same problem, because it calls internaly `setLatLngs`:
https://github.com/Leaflet/Leaflet/blob/620bc3e9c5d282090f429ac21d28d00fbffdb3db/src/layer/vector/Rectangle.js#L38-L40

So when I call `rect.setBounds([[0,0],[0,0]])` it also removes the last latlng, because it equals `[0,0] == [0,0]`

**Comment by johnd0e** [comment]
> * Default creation over Bounds L.rectangle([[0,0],[1,1]])

Yes, that really can be our starting point - rectange is defined by it's bounds.

> * Minimum / Maximum 4 LatLngs?

I cannot tell anything about number of latlngs. After some refactoring there may be even no latlngs stored!


> * Are rectangles always horizontal? This would make problems with some Plugins. For example with rotation.

Are you able to set non-horizontal Bounds?
I can imagine that you need some rotated rectangles, but it is not purpose of L.Rectangle, by it's definition.

> So when I call `rect.setBounds([[0,0],[0,0]])`

I don't know why do you need such zero-rectangle, but even if internally it will be represented by single point - it still be not against documentation, while `rect.getBounds()` returns same `[[0,0],[0,0]])`.
But I personally haven't checked all possible edge cases, and it's possible that current implementation may fail under certain conditions (feel free to report).

As for your issue: I suggest you extend `L.Rectangle` to add any methods/properties required for your goals.
It's also possible that your solution will be good enough to share with community as plugin.
Or merged into core (in case if it will be robust enough to cover all edge cases).

**Comment by Falke-Design** [comment]
> Are you able to set non-horizontal Bounds?

No I'm not, because of this the method `setLatLngs()` is needed. 

> I can imagine that you need some rotated rectangles, but it is not purpose of L.Rectangle, by it's definition.

I don't think that it would be the right / correct way to say that rectangles can only horizontal Bounds. What would then the differences between Bounds and Rectangles (except the visualizing). No latlngs, only bounds can be set / get and is always horizontal. In my eyes then would it better to name it BoundsRectangle or something that make it clear that it is "connected" to bounds.

Math definition of a Rectangle:
- has 4 sides / corners / angles
- the sides are parallel
- the length of opposide sides are equal
- all angles are 90°

So there is no definition that it must be horizontal. [Src 1](https://www.mathsisfun.com/geometry/rectangle.html) [Src 2](https://mathblog.com/reference/geometry/rectangle/)

I understand when we say that leaflet don't support rotated rectangles but I think it should be possible that it can be set from outside for example with `setLatLngs()`. (By the way this is a nice method to create rotated rectangles with two opposite corners (like bounds) [_getRotatedRectangle](https://github.com/geoman-io/leaflet-geoman/blob/7d9a8cf8423a989b7bd52711f350ca4fcbe5986e/src/js/L.PM.Utils.js#L160) )

> while rect.getBounds() returns same [[0,0],[0,0]])

Ok I don't tried it with `getBounds()` ... Then it is correct, but with `getLatLngs()` only 3 latlngs are returned.

> I don't know why do you need such zero-rectangle

This is a result of a editing function. When the user resizes the rectangle and drag the corners on top of each other, it can result in ` [[0,0],[0,0]]` 


----

I see (or it looks like) that you don't really understand me and don't think that it is a bug that it only hat 3 latlngs, but a rectangle is existing out of bounds, bounds are out of 4 latlngs. So why should it valid to remove the last latlng, when it is equal to the first one and then it has only 3 latlngs anymore. 
So I suggest to add this to L.Rectangle:
```
 _convertLatLngs: function (latlngs) { 
 	var result = Polyline.prototype._convertLatLngs.call(this, latlngs), 
 	    len = result.length; 
  
 	// remove last point if it equals first one 
 	if (len >= 4 && result[0] instanceof LatLng && result[0].equals(result[len - 1])) { 
 		result.pop(); 
 	} 
 	return result; 
 } 
```
No big change and it is never possible that leaflet destroys internally the rectangle latlngs. (It is something else if the user only pass 3 latlngs to `setLatLng`)

And in future maybe it has a better defenition but for now I think it is a bug ...

**Comment by johnd0e** [comment]
> I see (or it looks like) that you don't really understand me and don't think that it is a bug that it only hat 3 latlngs, but a rectangle is existing out of bounds, bounds are out of 4 latlngs. So why should it valid to remove the last latlng, when it is equal to the first one and then it has only 3 latlngs anymore.

Sure I understand you.
You are making use of operating latlngs directly, and found some case that is not handled well in your way.
While it's very easy to solve your issue locally, I'm still not convinced that it will be useful for Leaflet in general.
It's obvious that Leaflet's Rectangle is very limited and not even close to it's math definition.

And we cannot just take SOME properties from that definition ignoring all others.
And it's even worse Leaflet shapes are screen only, not geodesic, so they keep their properties only in current projection.

Anyway, I am keeping this issue open, may be other maintainers have own opinions.
